File: ParallelBatchEval/Julia/bench.jl
include("./FSP.jl")
include("./QAP.jl")
include("./Q3AP.jl")
include("./TSP.jl")

using Random
using Base.Threads

using .FSP
using .QAP
using .Q3AP
using .TSP

mutable struct Solution
    perm::Array{Int}
    cost::Int
end

mutable struct Solution2
    perm1::Array{Int,1}
    perm2::Array{Int,1}
    cost::Int
end

##################################### "constructors"
function randSol(n)
    sol=Solution(randperm(n),0)
end

function randSol2(n)
    sol=Solution2(randperm(n),randperm(n),0)
end

function FSPbatch(pop,PTM,nbJob,nbMach)
    t=@elapsed Threads.@threads for s in pop
        s.cost = evalMakespan(s.perm,PTM,nbJob,nbMach)
    end
    return t
end

function QAPbatch(pop,dist,flow,dim)
    t=@elapsed Threads.@threads for s in pop
        s.cost = evaluate_QAP(s.perm,dist,flow,dim)
    end
    return t
end

function Q3APbatch(pop,C6,dim)
    t=@elapsed Threads.@threads for s in pop
        s.cost = evaluate_Q3APP(s.perm1,s.perm2,C6,dim)
    end
    return t
end

function TSPbatch(pop,xcoord,ycoord,dim)
    t=@elapsed Threads.@threads for s in pop
        c = evaluate_TSP(s.perm,xcoord,ycoord,dim)
    end
    return t
end

###########################################################
if ARGS[1]=="FSP"
###########################################################
    println("flowshop --> instance? (ta001,...,ta120) : ",ARGS[2])
    read_in_FSP(ARGS[2])
    #warmup
    a=randSol(nbJob)
    println("first eval:")
    @time cmax1=evalMakespan(a.perm,PTM,nbJob,nbMach)
    println("second eval:")
    a=randSol(nbJob)
    @time cmax1=evalMakespan(a.perm,PTM,nbJob,nbMach)

    #read batchsize
    batchsize=parse(Int,ARGS[3])
    repeat=1e6/batchsize

    pop=[randSol(nbJob) for x in 1:batchsize]
    testFSP(pop,PTM,nbJob,nbMach)

    evaltime=0
    elap = @elapsed for i in 1:repeat
        pop=[randSol(nbJob) for x in 1:batchsize]
        global evaltime += FSPbatch(pop,PTM,nbJob,nbMach)
    end
    println("EvalLoop(s)\t",evaltime)
    println("TotElapsed(s)\t",elap)
###########################################################
elseif ARGS[1]=="QAP"
###########################################################
    println("QAP --> instance? (nug12,tho150,...) : ",ARGS[2])
    read_in_QAP(ARGS[2])
    #warmup
    a=randSol(QAP.dim)
    println("first eval:")
    @time a.cost=evaluate_QAP(a.perm,dist,flow,QAP.dim)
    println("second eval:")
    a=randSol(QAP.dim)
    @time a.cost=evaluate_QAP(a.perm,dist,flow,QAP.dim)

    #read batchsize
    batchsize=parse(Int,ARGS[3])
    repeat=1e6/batchsize
    #warmup
    pop=[randSol(QAP.dim) for x in 1:batchsize]
    QAPbatch(pop,dist,flow,QAP.dim)

    evaltime=0
    elap = @elapsed for i in 1:repeat
        pop=[randSol(QAP.dim) for x in 1:batchsize]
        global evaltime += QAPbatch(pop,dist,flow,QAP.dim)
    end
    println("EvalLoop(s)\t",evaltime)
    println("TotElapsed(s)\t",elap)
###########################################################
elseif ARGS[1]=="Q3AP"
###########################################################
    println("Q3AP --> instance? (nug12,...,nug25) : ",ARGS[2])
    read_in_Q3AP(ARGS[2])

    #warmup
    a=randSol2(Q3AP.dim)
    println("first eval:")
    @time a.cost=evaluate_Q3AP(a,C6,Q3AP.dim)
    println("second eval:")
    a=randSol2(Q3AP.dim)
    @time a.cost=evaluate_Q3AP(a,C6,Q3AP.dim)

    #read batchsize
    batchsize=parse(Int,ARGS[3])
    repeat=1e6/batchsize

    pop=[randSol2(Q3AP.dim) for x in 1:batchsize]
    Q3APbatch(pop,C6,Q3AP.dim)

    evaltime=0
    elap = @elapsed for i in 1:repeat
        pop=[randSol2(Q3AP.dim) for x in 1:batchsize]
        global evaltime += Q3APbatch(pop,C6,Q3AP.dim)
    end
    println("EvalLoop(s)\t",evaltime)
    println("TotElapsed(s)\t",elap)
###########################################################
elseif ARGS[1]=="TSP"
    println("Q3AP --> instance? (berlin52,...,) : ",ARGS[2])
    read_in_TSP(ARGS[2])

    #warmup
    a=randSol(TSP.dim)
    println("first eval:")
    @time c=evaluate_TSP(a.perm,xcoord,ycoord,TSP.dim)
    println("second eval:")
    a=randSol(TSP.dim)
    @time c=evaluate_TSP(a.perm,xcoord,ycoord,TSP.dim)

    #read batchsize
    batchsize=parse(Int,ARGS[3])
    repeat=1e6/batchsize

    #warmup
    pop=[randSol(TSP.dim) for x in 1:batchsize]
    TSPbatch(pop,xcoord,ycoord,TSP.dim)

    evaltime=0
    elap = @elapsed for i in 1:repeat
        pop=[randSol(TSP.dim) for x in 1:batchsize]
        global evaltime += TSPbatch(pop,xcoord,ycoord,TSP.dim)
    end
    println("EvalLoop(s)\t",evaltime)
    println("TotElapsed(s)\t",elap)
end

File: ParallelBatchEval/Py/bench.py
import sys
import time
import math
import numpy as np

from numba import jitclass, prange
from numba import int64, njit, threading_layer

from numba.typed import List

import Q3AP
import QAP
import TSP
import FSP

spec = [
    ('perm1', int64[:]),
    ('perm2', int64[:]),
    ('cost', int64),
]

@jitclass(spec)
class solution(object):
    def __init__(self, perm1, perm2, cost):
        self.perm1 = perm1
        self.perm2 = perm2
        self.cost = cost

    def solcopy(self):
        return solution(self.perm1.copy(), self.perm2.copy(), self.cost)

#=====================================
spec = [
    ('perm', int64[:]),
    ('cost', int64),
]

@jitclass(spec)
class solution1(object):
    def __init__(self, perm, cost):
        self.perm = perm
        self.cost = cost

    def solcopy(self):
        return solution(self.perm.copy(), self.cost)

########################################################
#Evaluation functions
########################################################
#EVAL Q3AP
@njit
def evalQ3AP(perm1,perm2):
    cost = 0
    for i in range(dim):
        for j in range(dim):
            cost += C[i, perm1[i], perm2[i],
                      j, perm1[j], perm2[j]]
    return cost

#EVAL QAP
@njit
def evalQAP(perm):
    cost = 0
    for i in range(dim):
        for j in range(dim):
            cost += dist[i,j]*flow[perm[i], perm[j]]
    return cost


#EVAL FSP
@njit
def evalFSP(perm):
    tmp = np.zeros(nbMach,dtype=int64)

    for i in range(nbJob):
        jb = perm[i]
        tmp[0] += PTM[0,jb]
        for j in range(1,nbMach):
            tmp[j] = max(tmp[j],tmp[j-1]) + PTM[j,jb]

    return tmp[nbMach-1]


@njit
def getCost(a, b):
    dx = xcoord[a] - xcoord[b]
    dy = ycoord[a] - ycoord[b]
    return dx * dx + dy * dy


@njit
def evalTSP(perm):
	# global dim
	cost = math.sqrt(getCost(0, perm[1]))
	for i in range(1,dim-1):
		cost += math.sqrt(getCost(perm[i], perm[i + 1]))
	cost += math.sqrt(getCost(perm[dim - 1], 0))
	return cost
########################################################

########################################################
#Parallel Batch Evaluation functions
########################################################
@njit(parallel=True)
def testQAP(pop):
    costs = np.zeros(batchsize)
    for i in prange(batchsize):
        costs[i] = evalQAP(pop[i]['perm'])
    return costs

#fails!
# @njit(parallel=True)
# def testQAPP(lst):
#     for i in prange(batchsize):
#         c=evalQAP(lst[i])

# @njit
# def popGen(size,batchsize):
#     lst = List()
#     for i in range(batchsize):
#         lst.append(np.random.permutation(size))
#     return lst

@njit(parallel=True)
def testFSP(pop):
    costs = np.zeros(batchsize)
    for i in prange(batchsize):
        costs[i] = evalFSP(pop[i]['perm'])
    # print(costs)
    return costs

@njit(parallel=True)
def testTSP(pop):
    costs = np.zeros(batchsize)
    for i in prange(batchsize):
        costs[i] = evalTSP(pop[i]['perm'])
    return costs

@njit(parallel=True)
def testQ3AP(pop):
    costs = np.zeros(batchsize)
    for i in prange(batchsize):
        costs[i] = evalQ3AP(pop[i]['perm1'],pop[i]['perm2'])
    return costs

########################################################
batchsize = int(sys.argv[3])
repeat = int(1e6/batchsize)

####################################
#read input
if sys.argv[1] == "QAP":
    print("QAP")
    dist,flow,dim = QAP.read_input(sys.argv[2])
    # print(dim,dist,flow)

    sol = solution1(np.arange(dim, dtype=np.int64), 0)
    sol.cost = evalQAP(sol.perm)
    print(sol.cost)

    sol1_type = np.dtype([('perm', np.int64, (dim,)), ('cost', np.int64, (1,))])
    pop = np.array([(np.random.permutation(dim), 0) for x in range(batchsize)], dtype=sol1_type)
    testQAP(pop)

    t1start = time.time()
    t2elapsed = 0.0

    for i in range(repeat):
        pop = np.array([(np.random.permutation(dim), 0) for x in range(batchsize)], dtype=sol1_type)

        t2start = time.time()
        testQAP(pop)
        t2elapsed += time.time() - t2start

    elapsed_time = time.time() - t1start
    print('EvalLoop:\t', t2elapsed)
    print('Elapsed:\t', elapsed_time)

###########################################################
if sys.argv[1] == "Q3AP":
###########################################################
    print("Q3AP - instance? (nug12,...): ",sys.argv[2])
    C,dim = Q3AP.read_input(sys.argv[2])
	#warmup
    sol = solution(np.arange(dim, dtype=np.int64), np.arange(dim, dtype=np.int64), 0)
    sol.cost = evalQ3AP(sol.perm1,sol.perm2)

	#needed....
    sol_type = np.dtype([('perm1', np.int64, (dim,)), ('perm2', np.int64, (dim,)), ('cost', np.int64, (1,))])
    pop = np.array([(np.random.permutation(dim), np.random.permutation(dim), 0) for x in range(batchsize)], dtype=sol_type)
    testQ3AP(pop)

    t1start = time.time()
    t2elapsed = 0.0
    for i in range(repeat):
        pop = np.array([(np.random.permutation(dim), np.random.permutation(dim), 0) for x in range(batchsize)], dtype=sol_type)
        t2start = time.time()
        testQ3AP(pop)
        t2elapsed += time.time() - t2start

    elapsed_time = time.time() - t1start
    print('EvalLoop:\t', t2elapsed)
    print('Elapsed:\t', elapsed_time)

###########################################################
if sys.argv[1] == "TSP":
###########################################################
    print("TSP - instance? (berlin52,...): ",sys.argv[2])
    xcoord,ycoord,dim = TSP.read_input(sys.argv[2])
	#warmup
    sol = solution1(np.arange(dim, dtype=np.int64), 0)
    sol.cost = evalTSP(sol.perm)

    sol1_type = np.dtype([('perm', np.int64, (dim,)), ('cost', np.int64, (1,))])
    pop = np.array([(np.random.permutation(dim), 0) for x in range(batchsize)], dtype=sol1_type)
    testTSP(pop)

    t1start = time.time()
    t2elapsed = 0.0
    for i in range(repeat):
        pop = np.array([(np.random.permutation(dim), 0) for x in range(batchsize)], dtype=sol1_type)
        t2start = time.time()
        testTSP(pop)
        t2elapsed += time.time() - t2start

    elapsed_time = time.time() - t1start
    print('EvalLoop:\t', t2elapsed)
    print('Elapsed:\t', elapsed_time)

###########################################################
if sys.argv[1] == "FSP":
###########################################################
    print("FSP - instance? (ta001,...): ",sys.argv[2])
    PTM,nbJob,nbMach = FSP.read_input(sys.argv[2])
	#warmup
    sol = solution1(np.arange(nbJob, dtype=np.int64), 0)
    sol.cost = evalFSP(sol.perm)

    sol1_type = np.dtype([('perm', np.int64, (nbJob,)), ('cost', np.int64, (1,))])
    pop = np.array([(np.random.permutation(nbJob), 0) for x in range(batchsize)], dtype=sol1_type)
    testFSP(pop)

    t1start = time.time()
    t2elapsed = 0.0
    for i in range(repeat):
        pop = np.array([(np.random.permutation(nbJob), 0) for x in range(batchsize)], dtype=sol1_type)
        t2start = time.time()
        testFSP(pop)
        t2elapsed += time.time() - t2start

    elapsed_time = time.time() - t1start
    print('EvalLoop:\t', t2elapsed)
    print('Elapsed:\t', elapsed_time)
